test/e2e/utils/helpers.go

```diff
@@ -160,48 +161,16 @@ func IsScyllaClusterRolledOut(sc *scyllav1.ScyllaCluster) (bool, error) {
 	return true, nil
 }
 
-func WaitForScyllaClusterState(ctx context.Context, client scyllav1client.ScyllaV1Interface, namespace string, name string, conditions ...func(sc *scyllav1.ScyllaCluster) (bool, error)) (*scyllav1.ScyllaCluster, error) {
-	fieldSelector := fields.OneTermEqualSelector("metadata.name", name).String()
-	lw := &cache.ListWatch{
-		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
-			options.FieldSelector = fieldSelector
-			return client.ScyllaClusters(namespace).List(ctx, options)
-		},
-		WatchFunc: func(options metav1.ListOptions) (i watch.Interface, e error) {
-			options.FieldSelector = fieldSelector
-			return client.ScyllaClusters(namespace).Watch(ctx, options)
-		},
-	}
-
-	event, err := watchtools.UntilWithSync(ctx, lw, &scyllav1.ScyllaCluster{}, nil, func(event watch.Event) (bool, error) {
-		switch event.Type {
-		case watch.Added, watch.Modified:
-			for _, condition := range conditions {
-				done, err := condition(event.Object.(*scyllav1.ScyllaCluster))
-				if err != nil {
-					return false, err
-				}
-				if !done {
-					return false, nil
-				}
-			}
-			return true, nil
-		default:
-			return true, fmt.Errorf("unexpected event: %#v", event)
-		}
-	})
-	if err != nil {
-		return nil, err
-	}
-
-	return event.Object.(*scyllav1.ScyllaCluster), nil
+type listerWatcher[ListPtr runtime.Object] interface {
+	List(context.Context, metav1.ListOptions) (ListPtr, error)
+	Watch(context.Context, metav1.ListOptions) (watch.Interface, error)
 }
 
 type WaitForStateOptions struct {
 	TolerateDelete bool
 }
 
-func WaitForPodState(ctx context.Context, client corev1client.PodInterface, name string, condition func(sc *corev1.Pod) (bool, error), o WaitForStateOptions) (*corev1.Pod, error) {
+func WaitForObjectState[ObjectPtrType, ObjectListPtrType runtime.Object](ctx context.Context, client listerWatcher[ObjectListPtrType], name string, options WaitForStateOptions, condition func(obj ObjectPtrType) (bool, error), additionalConditions ...func(obj ObjectPtrType) (bool, error)) (ObjectPtrType, error) {
```

Comment:
```suggestion
func WaitForObjectState[ObjectType, ObjectListType runtime.Object](ctx context.Context, client listerWatcher[ObjectListType], name string, options WaitForStateOptions, condition func(obj ObjectType) (bool, error), additionalConditions ...func(obj ObjectType) (bool, error)) (ObjectType, error) {
```

Reply:
It's on purpose as it isn't the ObjectType but a pointer to it (ObjectType is `corev1.Pod` but ObjectPtrType is `*corev1.Pod`)

Reply:
I really wanted to get the non-pointer types as arguments but without the constraints, golang didn't let me cast the interfaces later on. Good old c++ templates...

Reply:
> It's on purpose as it isn't the ObjectType but a pointer to it (ObjectType is corev1.Pod but ObjectPtrType is *corev1.Pod)

Name doesn't infer any condition on type, constraint is to implement runtime.Object, not being a pointer. It's just that `corev1.Pod` doesn't implement it, but `*corev1.Pod` does.

Reply:
`Type` seems redundant in this scope, so I would be fine with both `WaitForObjectState[Object, ObjectList runtime.Object]` and suggested one.

Reply:
done